lib/raty.js

```diff
@@ -0,0 +1,784 @@
+
+    for (var i = 0; i < options.length; i++) {
+      if (typeof this.opt[options[i]] === 'function') {
+        var value = this.opt[options[i]].call(this.element);
+
+        if (value) {
+          this.opt[options[i]] = value;
+        } else {
+          delete this.opt[options[i]];
+        }
+      }
+    }
+  }
+
+  // TODO: model spec
+  _fill(score) {
+    var hash = 0;
+
+    if (this.opt.iconRangeSame && this.opt.iconRange) {
+      while (hash < this.opt.iconRange.length && this.opt.iconRange[hash].range < score) {
+        hash++;
+      }
+    }
+
+    for (var i = 1; i <= this.stars.length; i++) {
+      var icon;
+      var star = this.stars[i - 1];
+      var turnOn = this._turnOn(i, score);
+
+      if (this.opt.iconRange && this.opt.iconRange.length > hash) {
+        var irange = this.opt.iconRange[hash];
+
+        icon = this._getRangeIcon(irange, turnOn);
+
+        if (i <= irange.range) {
+          this._setIcon(star, icon);
+        }
+
+        if (i === irange.range) {
+          hash++;
+        }
+      } else {
+        icon = this.opt[turnOn ? 'starOn' : 'starOff'];
+
+        this._setIcon(star, icon);
+      }
+    }
+  }
+
+  _getDecimal(number, fractions) {
+    var decimal = number.toString().split('.')[1];
+    var result = 0;
+
+    if (decimal) {
+      result = parseInt(decimal.slice(0, fractions), 10);
+
+      if (decimal.slice(1, 5) === '9999') {
+        result++;
+      }
+    }
+
+    return result;
+  }
+
+  // TODO: model spec
+  _getRangeIcon(irange, turnOn) {
+    return turnOn ? irange.on || this.opt.starOn : irange.off || this.opt.starOff;
+  }
+
+  // TODO: model spec
+  _getScoreByPosition(evt, icon) {
+    var score = parseInt(icon.alt || icon.getAttribute('data-alt'), 10);
+    if (this.opt.half) {
+      var size = this._getWidth();
+      var percent = parseFloat((evt.pageX - icon.offsetLeft) / size);
+      score = score - 1 + percent;
+    }
+    return score;
+  }
+
+  // TODO: model spec
+  _getHint(score, evt) {
+    if (score !== 0 && !score) {
+      return this.opt.noRatedMsg;
+    }
+
+    var decimal = this._getDecimal(score, 1);
+    var integer = Math.ceil(score);
+    var group = this.opt.hints[(integer || 1) - 1];
+    var hint = group;
+    var set = !evt || this.isMove;
+
+    if (this.opt.precision) {
+      if (set) {
+        decimal = decimal === 0 ? 9 : decimal - 1;
+      }
+
+      hint = group[decimal];
+    } else if (this.opt.halfShow || this.opt.half) {
+      decimal = set && decimal === 0 ? 1 : decimal > 5 ? 1 : 0;
+
+      hint = group[decimal];
+    }
+
+    return hint === '' ? '' : hint || score;
+  }
+
+  // TODO: model spec
+  _getWidth() {
+    var width = this.stars[0].width || parseFloat(this.stars[0].offsetWidth);
+    if (!width) {
+      this._error('Could not get the icon width!');
+    }
+
+    return width;
+  }
+
+  // TODO: model spec
+  _lock() {
+    var hint = this._getHint(this?.scoreField?.value);
```

Comment:
Why we're using save navigation? The score field should always be present, no? Before we hadn't it.